### PR TITLE
[foreach] Speed up slow path perf for foreach binary ops

### PR DESCRIPTION
**Pull request by janeyx99**
For the slow path scalar overload, we for-loop call aten::op(tensor_from_list, scalar). This is inefficient as it will call empty_strided and copy_ for every iter of the loop. 

We can likely take advantage of the heuristics and make the tensor from the scalar only once. The main benefit of this change will be for optimizers, where foreach_add_(steps, 1) is a common codepath and the step tensors are on CPU.

Anticipated qs:
1. what if the scalarlist is empty? we error on this earlier in the stack, so the kernel can assume the tensorlist has content.
2. what if the rest of the list doesn't match the first tensor in shape, device, dtype, layout? Broadcasting solves for shape and layout. We add a check for dtype + device to maintain BCness. 
3. why do I guard on `is_floating_point()`? If we create a bool tensor, the wrong behavior happens where `torch.tensor(True) + 3` will return `torch.tensor(True)` instead of `torch.tensor(4)`. IMO the first behavior makes more sense but would be BC breaking.
4. why not use `wrapped_scalar_tensor()`? We could. However, it defaults to kDouble instead of kFloat due to Python scalars being 64bits wide, which is inconvenient since most training is done in lower precision. Thus, an extra `.to(dtype)` would be required at the very least, and there would still need to be some sort of check added to ensure correctness. This, using `wrapped_scalar_tensor()` did not seem more performant nor clearer.

Code to test:
```
import torch
steps = [torch.zeros((), device="cpu") for i in range(1000)]

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    torch._foreach_add_(steps, 1024.1024)

print(p.key_averages().table(sort_by="cpu_time_total"))
```

Profile (for 1000 tensors in a list):

```
current state of world on main:
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
       aten::_foreach_add        37.43%      52.961ms        99.99%     141.461ms     141.461ms             1  
                aten::add        28.48%      40.300ms        62.55%      88.500ms      88.500us          1000  
                 aten::to         4.12%       5.825ms        34.07%      48.200ms      48.200us          1000  
           aten::_to_copy        10.67%      15.102ms        29.95%      42.375ms      42.375us          1000  
              aten::copy_        14.92%      21.103ms        14.92%      21.103ms      21.103us          1000  
      aten::empty_strided         4.36%       6.170ms         4.36%       6.170ms       6.170us          1000  
    cudaDeviceSynchronize         0.01%      19.000us         0.01%      19.000us      19.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 141.480ms

new state after pr:
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        74.60%      53.748ms        99.97%      72.022ms      72.022ms             1  
               aten::add_        25.28%      18.212ms        25.28%      18.212ms      18.212us          1000  
          aten::new_empty         0.03%      22.000us         0.07%      51.000us      51.000us             1  
              aten::empty         0.04%      29.000us         0.04%      29.000us      29.000us             1  
    cudaDeviceSynchronize         0.03%      22.000us         0.03%      22.000us      22.000us             1  
              aten::fill_         0.02%      11.000us         0.02%      11.000us      11.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 72.044ms
```
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110954

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by janeyx99** [comment]
I am abandoning this change because type promotion is a real concern. Since binary ops are a broader category than just our happy add/sub/mul/div, I don't want to make assumptions about what the Scalar _should_ promote to too early. e.g., something like `eq` will promote types differently than `add`.

Anyway, the approach I will take instead is:
1. Add a scalar Tensor overload for _foreach_add
2. Change the call sites within our core optimizers to be more performant
